lib/knapsack_pro/runners/queue/rspec_runner.rb

```diff
@@ -3,224 +3,283 @@
 module KnapsackPro
   module Runners
     module Queue
+      # Imperative Shell [instance of KnapsackPro::Runners::Queue::RSpecRunner]
+      # Architecture based on: https://www.destroyallsoftware.com/talks/boundaries
+      #
+      # It should contain calls to RSpec methods.
+      # Preferably via methods with the  `knapsack__` prefix.
+      # We want to isolate the imperative shell from the internals of RSpec.
+      # It should be tested via integration tests if possible.
+      # Alternatively, it should be tested using E2E tests on CI that are running against a Rails app with the RSpec test suite.
       class RSpecRunner < BaseRunner
-        @@used_seed = nil
+        # Functional Core.
+        # Architecture based on: https://www.destroyallsoftware.com/talks/boundaries
+        #
+        # It has business logic related to how Knapsack Pro works.
+        # It should be easy to unit test it.
+        # It should NOT contain direct calls to RSpec.
+        class FunctionalCore
+          ADAPTER_CLASS = KnapsackPro::Adapters::RSpecAdapter
+          FAILURE_EXIT_CODE = 1
+          FORMATTERS = [
+            'KnapsackPro::Formatters::TimeTracker',
+          ]
 
-        def self.run(args)
-          require 'rspec/core'
-          require_relative '../../formatters/time_tracker'
-          require_relative '../../formatters/time_tracker_fetcher'
-          require_relative '../../formatters/rspec_queue_summary_formatter'
-          require_relative '../../formatters/rspec_queue_profile_formatter_extension'
+          def initialize(logger)
+            @logger = logger
+          end
 
-          ENV['KNAPSACK_PRO_TEST_SUITE_TOKEN'] = KnapsackPro::Config::Env.test_suite_token_rspec
-          ENV['KNAPSACK_PRO_QUEUE_RECORDING_ENABLED'] = 'true'
-          ENV['KNAPSACK_PRO_QUEUE_ID'] = KnapsackPro::Config::EnvGenerator.set_queue_id
+          def ensure_no_deprecated_run_all_when_everything_filtered_option!(deprecated_run_all_when_everything_filtered_enabled)
+            return unless deprecated_run_all_when_everything_filtered_enabled
 
-          KnapsackPro::Config::Env.set_test_runner_adapter(adapter_class)
-          runner = new(adapter_class)
+            error_message = "The run_all_when_everything_filtered option is deprecated. See: #{KnapsackPro::Urls::RSPEC__DEPRECATED_RUN_ALL_WHEN_EVERYTHING_FILTERED}"
+            logger.error(error_message)
+            raise error_message
```

Comment:
Side effect, shouldn't be in core.

Reply:
Is just raising a side effect, or is writing to a logger also a side effect? I guess both. Perhaps the whole method should be outside of the core.

Reply:
I guess the way you can summarize the Boundary architecture for the core is: you can only have methods/functions that you test with 100% coverage by passing an input and asserting on the output with no stubs or mocks.

You cannot test `raise` as on output, so it cannot be in core. And you cannot test `logger` without a stub/mock, so it cannot be in core.

Though, we don't have to adhere 100% to any architecture. Let's do so if it makes our lives easier.

In a way, I think that raising and logging outside of core does make it easier because those side effects makes more sense if they live in the shell with the other side effects (eg, raise and rescue closer together).

But then again, you are the author here 😉

Reply:
change: https://github.com/KnapsackPro/knapsack_pro-ruby/pull/237/commits/a4ccd512c12db90d5cdc9073e00bdac1b2a0b73d

I moved the `ensure_no_deprecated_run_all_when_everything_filtered_option!` method outside of the core. It's a private method now. It's tested in integration tests. We don't need a unit test.

Reply:
I did refactor to get rid of logger from the core for other methods too.